File: simulations/pred_mgn_est.R
#!/usr/bin/env Rscript
########################################
## input configurations
########################################
args <- commandArgs(trailingOnly = TRUE)
p <- as.integer(args[1])
n <- as.integer(args[2])
alpha_ind <- as.integer(args[3])
Gamma_ind <- as.integer(args[4])
seed <- as.integer(args[5])

alphas = seq(0.1,0.9,by=0.1)
gammas = c(1.5,2,2.5,3,5)
# coverage target 1-alpha
alpha = alphas[alpha_ind] 
# confounding level Gamma
Gamma = gammas[Gamma_ind] 

########################################
## load libraries
########################################
suppressPackageStartupMessages(library(grf))
options(warn=-1)

########################################
## load util functions
########################################
source("../utils/util_ate.R")
data_type <- "continuous"
cat(paste(" - Running the script with marginally-valid algorithm and estimated bounds, alpha ", alpha, ", Gamma ",Gamma,
          ", n ", n, ", p ", p, ", seed ", seed, ", type ", data_type, "\n"), sep = '')

########################################
## Output direcroty
########################################
if(!dir.exists("../results")){
  dir.create("../results")
}
out_dir <- "../results/simulation/"
if(!dir.exists(out_dir)){
  dir.create(out_dir)
}

########################################
## Parameter
########################################
alpha0 = 0
n_test = 500
beta = matrix(c(-0.531,0.126,-0.312,0.018,rep(0,p-4)), nrow=p)
noise_level = 0.5
# generate true probability of treatment
pp = mean(data.gen.ate(n*1000, p, Gamma, beta, alpha0, obs=FALSE)$T)


set.seed(seed)
########################################
## fit on the training fold
########################################
train.data = data.gen.ate(n, p, Gamma, beta, alpha0, obs=TRUE)
train.X = (train.data$X[train.data$T==1,])[1:n,]
train.Y = (train.data$Y1[train.data$T==1])[1:n]
# train the nonconformity score function
train.score = conform.score(train.X, train.Y, "cqr", trained_model=NULL, quantile=1-alpha)
t.mdl = train.score$model
# estimate hat{e}(x)
hat.p = mean(train.data$T)
# 调试输出：检查训练数据
cat(" [debug] Training data check:\n")
cat("   hat.p =", hat.p, "\n")
cat("   train.data$T (first 10):", head(train.data$T, 10), "\n")
cat("   train.data$X (first row, first 5):", head(train.data$X[1,], 5), "\n")
e.model = regression_forest(train.data$X, train.data$T, num.threads = 1)
# 调试输出：检查 e.model 训练后的预测
train.ex.check = predict(e.model, newdata=head(train.data$X, 5))$predictions
cat("   e.model predictions (first 5):", train.ex.check, "\n")

########################################
## calibration 
########################################
calib.data = data.gen.ate(n, p, Gamma, beta, alpha0, obs=TRUE)
calib.X = (calib.data$X[calib.data$T==1,])[1:n,]
calib.Y = (calib.data$Y1[calib.data$T==1])[1:n]
calib.ex = predict(e.model, newdata=calib.X)$predictions  
n_calib = length(calib.Y)
# 调试输出：检查校准数据的 e(x) 预测
cat(" [debug] Calibration e(x) check:\n")
cat("   calib.ex (first 5):", head(calib.ex, 5), "\n")
cat("   calib.ex (min, max, mean):", min(calib.ex), max(calib.ex), mean(calib.ex), "\n")

# 调试输出：检查 e(x) 的分布，看是否极端化
cat(" [debug] RF e(x) distribution:\n")
cat("   min:", min(calib.ex), ", max:", max(calib.ex), "\n")
cat("   mean:", mean(calib.ex), ", sd:", sd(calib.ex), "\n")
cat("   < 0.1:", sum(calib.ex < 0.1), ", > 0.9:", sum(calib.ex > 0.9), "\n")
cat("   < 0.01:", sum(calib.ex < 0.01), ", > 0.99:", sum(calib.ex > 0.99), "\n")

# lower and upper bounds of weight function
calib.lx = hat.p * (1+ (1-calib.ex) / (calib.ex*Gamma))
calib.ux = hat.p * (1+ Gamma * (1-calib.ex) / (calib.ex))
calib.nc.w = hat.p / calib.ex

# 调试输出：检查权重是否成比例
cat(" [debug] Gamma =", Gamma, "\n")
cat(" [debug] First 5 weights comparison:\n")
cat("   calib.lx (head):", head(calib.lx, 5), "\n")
cat("   calib.ux (head):", head(calib.ux, 5), "\n")
cat("   calib.wx (head):", head(calib.nc.w, 5), "\n")
cat("   ratio lx/wx (head):", head(calib.lx/calib.nc.w, 5), "\n")
cat("   ratio lx/wx (mean):", mean(calib.lx/calib.nc.w), ", sd:", sd(calib.lx/calib.nc.w), "\n")

# non-conformity score
calib.score = conform.score(calib.X, calib.Y, "cqr", trained_model=t.mdl, quantile=1-alpha)$score
calib.all = data.frame("score"=calib.score, "lx"=calib.lx, "ux"=calib.ux, "wx"=calib.nc.w, "ex" = calib.ex)
calib.all = calib.all[order(calib.all$score),]
rownames(calib.all) = 1:dim(calib.all)[1]

# calculate estimation error and actual gap
calib.true.exu = calib.data$exu
calib.true.ex = calib.data$ex
calib.true.wx = pp / calib.true.exu
gap = (mean(pmax(calib.lx - calib.true.wx, 0)) + mean(pmax(calib.true.wx - calib.ux, 0)) + 1/nrow(calib.X) * mean( calib.true.wx * max(pmax(calib.true.wx - calib.ux, 0))) ) * max(1/calib.lx)
l_err = mean(abs(calib.lx - pp * (1 + (1-calib.true.ex)/(calib.true.ex*Gamma))))
u_err = mean(abs(calib.ux - pp * (1 + Gamma * (1-calib.true.ex)/calib.true.ex)))
l_inv = mean(1/calib.lx)

########################################
## generate test fold
########################################
test.data = data.gen.ate(n_test, p, Gamma, beta, alpha0, obs=FALSE)
test.X = test.data$X
test.Y1 = test.data$Y1
test.ex = predict(e.model, newdata=test.X)$predictions 
test.lx = hat.p*(1+ (1-test.ex)/(test.ex*Gamma))
test.ux = hat.p*(1+ Gamma* (1-test.ex)/(test.ex))
test.wx = hat.p / test.ex
# 调试输出：检查测试集权重
cat(" [debug] Test weights comparison:\n")
cat("   test.lx (head):", head(test.lx, 5), "\n")
cat("   test.ux (head):", head(test.ux, 5), "\n")
cat("   test.wx (head):", head(test.wx, 5), "\n")
cat("   ratio lx/wx (head):", head(test.lx/test.wx, 5), "\n")
cat("   ratio lx/wx (mean):", mean(test.lx/test.wx), ", sd:", sd(test.lx/test.wx), "\n")
test.pred = predict(t.mdl, test.X, quantile=c(alpha/2, 1-alpha/2)) 
# ✅ 兼容 grf 新旧版本
if (is.list(test.pred) && "predictions" %in% names(test.pred)) {
  test.pred <- test.pred$predictions
} 

########################################
## the confounding-aware algorithm  
########################################

cat(" - Computing the robust weighted conformal inference...")
  
# partial sums for confounding-aware
sum.num = rep(0,n_calib) # for numerator
sum.den = rep(0,n_calib) # for denominator
sum.num[1] = calib.all$lx[1]
sum.den[1] = calib.all$lx[1] + sum(calib.all$ux[2:n_calib])
for (k in 2:n_calib){
  sum.num[k] = sum.num[k-1] + calib.all$lx[k]
  sum.den[k] = sum.den[k-1] - calib.all$ux[k] + calib.all$lx[k] 
}

# confounding-aware prediction 
c.test.lo = rep(0,n_test)
c.test.hi = rep(0,n_test)
for (ii in 1:n_test){
  ratios = sum.num / (sum.den + test.ux[ii])
  kstar = min(which(ratios>1-alpha))
  v.kstar = calib.all$score[kstar]
  c.test.lo[ii] = test.pred[ii,1]-v.kstar
  c.test.hi[ii] = test.pred[ii,2]+v.kstar
}
# evaluate coverage on test data
c.cover = (c.test.lo <= test.Y1) * (c.test.hi >= test.Y1)

cat("Done.\n")


########################################
## the confounding-unaware algorithm  
########################################

cat(" - Computing the vanilla weighted conformal inference...")

nc.sum = rep(0,n_calib)
nc.sum[1] = calib.all$wx[1]
for (k in 2:n_calib){
  nc.sum[k] = nc.sum[k-1] + calib.all$wx[k]
}

nc.test.lo = rep(0,n_test)
nc.test.hi = rep(0,n_test)
nc.test.weight = hat.p / test.ex

for (ii in 1:n_test){
  nc.ratios = nc.sum / (nc.sum[n_calib] + nc.test.weight[ii])
  nc.kstar = min(which(nc.ratios>1-alpha))
  nc.v.kstar = calib.all$score[nc.kstar]
  nc.test.lo[ii] = test.pred[ii,1] - nc.v.kstar
  nc.test.hi[ii] = test.pred[ii,2] + nc.v.kstar
}
nc.cover = (nc.test.lo <= test.Y1) * (nc.test.hi >= test.Y1)

cat("Done.\n")

########################################
# output summary of test   
########################################
res = data.frame("c.cov" = mean(c.cover), "c.len" = mean(c.test.hi-c.test.lo),
                 "nc.cov" = mean(nc.cover), "nc.len" = mean(nc.test.hi-nc.test.lo), 
                 "gap" = gap, "l_err" = l_err, "u_err" = u_err, "l_inv" = l_inv,
                 "n" = n, "p" = p, "n_calib" = n_calib,
                 "gamma" = Gamma, "alpha" = alpha, "seed" = seed, 
                 "data_type" = data_type, "method" = "marginal_est")

save.path = paste(out_dir, "pred_marginal_est_p_",p,"_n_",n,"_alpha_",alpha_ind,"_gamma_",Gamma_ind,"_seed_",seed,"_",data_type,".csv",sep='')
write.csv(res, save.path)

File: simulations/pred_mgn_est.py
#!/usr/bin/env python3
"""
Python version of pred_mgn_est.R with Neural Network replacing Random Forest
Marginally-valid algorithm with estimated bounds
Enables GPU acceleration for faster training and inference
"""

import sys
import os
import numpy as np
import pandas as pd
import torch

# Add parent directory to path to import utils
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
from utils.util_ate import data_gen_ate, conform_score, RegressionForestNN


def main():
    ########################################
    ## Input configurations
    ########################################
    if len(sys.argv) < 6:
        print("Usage: python pred_mgn_est.py <p> <n> <alpha_ind> <Gamma_ind> <seed>")
        sys.exit(1)
    
    p = int(sys.argv[1])
    n = int(sys.argv[2])
    alpha_ind = int(sys.argv[3])
    Gamma_ind = int(sys.argv[4])
    seed = int(sys.argv[5])
    
    alphas = np.arange(0.1, 1.0, 0.1)
    gammas = np.array([1.5, 2, 2.5, 3, 5])
    
    # Coverage target 1-alpha
    alpha = alphas[alpha_ind - 1]  # R uses 1-indexing
    # Confounding level Gamma
    Gamma = gammas[Gamma_ind - 1]  # R uses 1-indexing
    
    data_type = "continuous"
    print(f" - Running the script with marginally-valid algorithm and estimated bounds, "
          f"alpha {alpha}, Gamma {Gamma}, n {n}, p {p}, seed {seed}, type {data_type}")
    
    ########################################
    ## Output directory
    ########################################
    out_dir = "../results/simulation/"
    os.makedirs(out_dir, exist_ok=True)
    
    ########################################
    ## Parameters
    ########################################
    alpha0 = 0
    n_test = 500
    beta_base = np.array([[-0.531], [0.126], [-0.312], [0.018]])
    if p > 4:
        beta = np.vstack([beta_base, np.zeros((p - 4, 1))])
    else:
        beta = beta_base[:p]
    
    noise_level = 0.5
    
    # Generate true probability of treatment
    temp_data = data_gen_ate(n * 1000, p, Gamma, beta, alpha0, obs=False)
    pp = np.mean(temp_data['T'])
    
    # Set seed for reproducibility (same as R version)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
    
    ########################################
    ## Fit on the training fold
    ########################################
    train_data = data_gen_ate(n, p, Gamma, beta, alpha0, obs=True)
    
    # Extract treated samples
    treated_idx = np.where(train_data['T'] == 1)[0][:n]
    train_X = train_data['X'][treated_idx]
    train_Y = train_data['Y1'][treated_idx]
    
    # Train the nonconformity score function (NN instead of quantile_forest)
    print(" - Training quantile regression neural network...")
    device = 'cuda'  # Use GPU
    train_score = conform_score(train_X, train_Y, method='cqr', trained_model=None,
                                quantile=1 - alpha, device=device, epochs=100, 
                                batch_size=32, verbose=False)
    t_mdl = train_score['model']
    
    # Estimate hat{e}(x) using RegressionForestNN (NN instead of regression_forest)
    hat_p = np.mean(train_data['T'])
    print(f" [debug] Training data check:")
    print(f"   hat.p = {hat_p}")
    print(f"   train.data['T'] (first 10): {train_data['T'][:10]}")
    print(f"   train.data['X'] (first row, first 5): {train_data['X'][0, :5]}")
    
    print(" - Training propensity score neural network...")
    e_model = RegressionForestNN(train_data['X'], train_data['T'], device=device,
                                 epochs=100, batch_size=32, verbose=False)
    
    # Check e_model predictions
    train_ex_check = e_model.predict(train_data['X'][:5])['predictions']
    print(f"   e.model predictions (first 5): {train_ex_check}")
    
    ########################################
    ## Calibration
    ########################################
    calib_data = data_gen_ate(n, p, Gamma, beta, alpha0, obs=True)
    
    # Extract treated samples for calibration
    calib_treated_idx = np.where(calib_data['T'] == 1)[0][:n]
    calib_X = calib_data['X'][calib_treated_idx]
    calib_Y = calib_data['Y1'][calib_treated_idx]
    
    # Predict e(x) for calibration set
    calib_ex = e_model.predict(calib_X)['predictions']
    n_calib = len(calib_Y)
    
    # Debug output: check calibration e(x) predictions
    print(f" [debug] Calibration e(x) check:")
    print(f"   calib.ex (first 5): {calib_ex[:5]}")
    print(f"   calib.ex (min, max, mean): {np.min(calib_ex)}, {np.max(calib_ex)}, {np.mean(calib_ex)}")
    
    # Debug output: check e(x) distribution
    print(f" [debug] NN e(x) distribution:")
    print(f"   min: {np.min(calib_ex)}, max: {np.max(calib_ex)}")
    print(f"   mean: {np.mean(calib_ex)}, sd: {np.std(calib_ex)}")
    print(f"   < 0.1: {np.sum(calib_ex < 0.1)}, > 0.9: {np.sum(calib_ex > 0.9)}")
    print(f"   < 0.01: {np.sum(calib_ex < 0.01)}, > 0.99: {np.sum(calib_ex > 0.99)}")
    
    # Lower and upper bounds of weight function
    calib_lx = hat_p * (1 + (1 - calib_ex) / (calib_ex * Gamma))
    calib_ux = hat_p * (1 + Gamma * (1 - calib_ex) / calib_ex)
    calib_nc_w = hat_p / calib_ex
    
    # Debug output: check weights
    print(f" [debug] Gamma = {Gamma}")
    print(f" [debug] First 5 weights comparison:")
    print(f"   calib.lx (head): {calib_lx[:5]}")
    print(f"   calib.ux (head): {calib_ux[:5]}")
    print(f"   calib.wx (head): {calib_nc_w[:5]}")
    print(f"   ratio lx/wx (head): {(calib_lx / calib_nc_w)[:5]}")
    print(f"   ratio lx/wx (mean): {np.mean(calib_lx / calib_nc_w)}, sd: {np.std(calib_lx / calib_nc_w)}")
    
    # Non-conformity score
    calib_score = conform_score(calib_X, calib_Y, method='cqr', trained_model=t_mdl,
                                quantile=1 - alpha, device=device)['score']
    
    # Create DataFrame and sort by score
    calib_all = pd.DataFrame({
        'score': calib_score,
        'lx': calib_lx,
        'ux': calib_ux,
        'wx': calib_nc_w,
        'ex': calib_ex
    })
    calib_all = calib_all.sort_values('score').reset_index(drop=True)
    
    # Calculate estimation error and actual gap
    calib_true_exu = calib_data['exu'][calib_treated_idx]
    calib_true_ex = calib_data['ex'][calib_treated_idx]
    calib_true_wx = pp / calib_true_exu
    
    gap = (np.mean(np.maximum(calib_lx - calib_true_wx, 0)) + 
           np.mean(np.maximum(calib_true_wx - calib_ux, 0)) + 
           1 / len(calib_X) * np.mean(calib_true_wx * np.max(np.maximum(calib_true_wx - calib_ux, 0)))) * np.max(1 / calib_lx)
    
    l_err = np.mean(np.abs(calib_lx - pp * (1 + (1 - calib_true_ex) / (calib_true_ex * Gamma))))
    u_err = np.mean(np.abs(calib_ux - pp * (1 + Gamma * (1 - calib_true_ex) / calib_true_ex)))
    l_inv = np.mean(1 / calib_lx)
    
    ########################################
    ## Generate test fold
    ########################################
    test_data = data_gen_ate(n_test, p, Gamma, beta, alpha0, obs=False)
    test_X = test_data['X']
    test_Y1 = test_data['Y1']
    
    # Predict e(x) for test set
    test_ex = e_model.predict(test_X)['predictions']
    test_lx = hat_p * (1 + (1 - test_ex) / (test_ex * Gamma))
    test_ux = hat_p * (1 + Gamma * (1 - test_ex) / test_ex)
    test_wx = hat_p / test_ex
    
    # Debug output: check test weights
    print(f" [debug] Test weights comparison:")
    print(f"   test.lx (head): {test_lx[:5]}")
    print(f"   test.ux (head): {test_ux[:5]}")
    print(f"   test.wx (head): {test_wx[:5]}")
    print(f"   ratio lx/wx (head): {(test_lx / test_wx)[:5]}")
    print(f"   ratio lx/wx (mean): {np.mean(test_lx / test_wx)}, sd: {np.std(test_lx / test_wx)}")
    
    # Predict quantiles
    test_pred = t_mdl.predict(test_X, quantile=[alpha / 2, 1 - alpha / 2])
    if isinstance(test_pred, dict) and 'predictions' in test_pred:
        test_pred = test_pred['predictions']
    
    ########################################
    ## The confounding-aware algorithm
    ########################################
    print(" - Computing the robust weighted conformal inference...", end="")
    
    # Partial sums for confounding-aware
    sum_num = np.zeros(n_calib)
    sum_den = np.zeros(n_calib)
    sum_num[0] = calib_all['lx'].values[0]
    sum_den[0] = calib_all['lx'].values[0] + np.sum(calib_all['ux'].values[1:])
    
    for k in range(1, n_calib):
        sum_num[k] = sum_num[k - 1] + calib_all['lx'].values[k]
        sum_den[k] = sum_den[k - 1] - calib_all['ux'].values[k] + calib_all['lx'].values[k]
    
    # Confounding-aware prediction
    c_test_lo = np.zeros(n_test)
    c_test_hi = np.zeros(n_test)
    
    for ii in range(n_test):
        ratios = sum_num / (sum_den + test_ux[ii])
        kstar = np.where(ratios > 1 - alpha)[0][0]  # First index where ratio > 1-alpha
        v_kstar = calib_all['score'].values[kstar]
        c_test_lo[ii] = test_pred[ii, 0] - v_kstar
        c_test_hi[ii] = test_pred[ii, 1] + v_kstar
    
    # Evaluate coverage on test data
    c_cover = (c_test_lo <= test_Y1) & (c_test_hi >= test_Y1)
    
    print("Done.")
    
    ########################################
    ## The confounding-unaware algorithm
    ########################################
    print(" - Computing the vanilla weighted conformal inference...", end="")
    
    nc_sum = np.zeros(n_calib)
    nc_sum[0] = calib_all['wx'].values[0]
    
    for k in range(1, n_calib):
        nc_sum[k] = nc_sum[k - 1] + calib_all['wx'].values[k]
    
    nc_test_lo = np.zeros(n_test)
    nc_test_hi = np.zeros(n_test)
    nc_test_weight = hat_p / test_ex
    
    for ii in range(n_test):
        nc_ratios = nc_sum / (nc_sum[n_calib - 1] + nc_test_weight[ii])
        nc_kstar = np.where(nc_ratios > 1 - alpha)[0][0]
        nc_v_kstar = calib_all['score'].values[nc_kstar]
        nc_test_lo[ii] = test_pred[ii, 0] - nc_v_kstar
        nc_test_hi[ii] = test_pred[ii, 1] + nc_v_kstar
    
    nc_cover = (nc_test_lo <= test_Y1) & (nc_test_hi >= test_Y1)
    
    print("Done.")
    
    ########################################
    ## Output summary of test
    ########################################
    res = pd.DataFrame({
        'c.cov': [np.mean(c_cover)],
        'c.len': [np.mean(c_test_hi - c_test_lo)],
        'nc.cov': [np.mean(nc_cover)],
        'nc.len': [np.mean(nc_test_hi - nc_test_lo)],
        'gap': [gap],
        'l_err': [l_err],
        'u_err': [u_err],
        'l_inv': [l_inv],
        'n': [n],
        'p': [p],
        'n_calib': [n_calib],
        'gamma': [Gamma],
        'alpha': [alpha],
        'seed': [seed],
        'data_type': [data_type],
        'method': ['marginal_est']
    })
    
    save_path = f"{out_dir}pred_marginal_est_p_{p}_n_{n}_alpha_{alpha_ind}_gamma_{Gamma_ind}_seed_{seed}_{data_type}_nn.csv"
    res.to_csv(save_path, index=False)
    
    print(f"\nResults saved to: {save_path}")
    print(f"  Confounding-aware coverage: {np.mean(c_cover):.4f}")
    print(f"  Confounding-unaware coverage: {np.mean(nc_cover):.4f}")
    print(f"  Confounding-aware interval length: {np.mean(c_test_hi - c_test_lo):.4f}")
    print(f"  Confounding-unaware interval length: {np.mean(nc_test_hi - nc_test_lo):.4f}")


if __name__ == "__main__":
    main()


